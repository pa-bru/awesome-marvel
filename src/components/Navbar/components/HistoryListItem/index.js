import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'

import IconButton from '@material-ui/core/IconButton'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction'
import Avatar from '@material-ui/core/Avatar'
import DeleteIcon from '@material-ui/icons/Delete'

export class HistoryListItem extends Component {
  static propTypes = {
    thumbnail: PropTypes.shape({
      path: PropTypes.string,
      extension: PropTypes.string,
    }),
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    series: PropTypes.shape({
      available: PropTypes.number,
    }).isRequired,
    comics: PropTypes.shape({
      available: PropTypes.number,
    }).isRequired,
    events: PropTypes.shape({
      available: PropTypes.number,
    }).isRequired,
    onClick: PropTypes.func.isRequired,
    history: PropTypes.object,
  }

  static defaultProps = {
    thumbnail: {
      path: '',
      extension: '',
    },
    series: {
      available: 0,
    },
    comics: {
      available: 0,
    },
    events: {
      available: 0,
    },
  }

  render() {
    const { thumbnail, id, name, comics, series, events, onClick, history } = this.props
    const secondary = `${comics.available} comics - ${series.available} series - ${
      events.available
    } events`

    return (
      <ListItem button dense onClick={() => history.push(`/characters/${id}`)}>
        <Avatar aria-label="Recipe" src={`${thumbnail.path}.${thumbnail.extension}`} />
        <ListItemText primary={name} secondary={secondary} />
        <ListItemSecondaryAction>
          <IconButton aria-label="Delete" onClick={onClick}>
            <DeleteIcon />
          </IconButton>
        </ListItemSecondaryAction>
      </ListItem>
    )
  }
}

export default withRouter(HistoryListItem)
