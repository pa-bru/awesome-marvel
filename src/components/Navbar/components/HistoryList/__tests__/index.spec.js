import React from 'react'
import { shallow } from 'enzyme'

import { HistoryList, mapStateToProps } from '../'

import HistoryListItem from '../../HistoryListItem'

const defaultProps = {
  loadHistoryList: jest.fn(),
  historyList: [],
  clearHistory: jest.fn(),
  removeFromHistory: jest.fn(),
  classes: {},
  history: {},
  isIncomplete: false,
}

describe('HistoryList', () => {
  it('should return the expected result of mapStateToProps', () => {
    // Given
    const store = {
      list: {
        history: {
          ids: [1, 2],
        },
      },
      heroes: {
        1: {
          id: 1,
        },
        2: {
          id: 2,
        },
      },
    }

    const expected = {
      isIncomplete: false,
      history: store.list.history,
      historyList: [
        {
          id: 1,
        },
        {
          id: 2,
        },
      ],
    }

    // When
    const state = mapStateToProps(store)

    // Then
    expect(state).toEqual(expected)
  })

  it('should render the error view', () => {
    // Given
    const props = {
      ...defaultProps,
      history: {
        error: {
          code: 400,
          status: 'Bad request',
        },
      },
    }
    // When
    const tree = shallow(<HistoryList {...props} />)

    // Then
    expect(tree).toMatchSnapshot()
  })

  it('should render the loading view', () => {
    // Given
    const props = {
      ...defaultProps,
      history: {
        isFetching: true,
      },
    }
    // When
    const tree = shallow(<HistoryList {...props} />)

    // Then
    expect(tree).toMatchSnapshot()
  })

  it('should render a specific message when the list is empty', () => {
    // When
    const tree = shallow(<HistoryList {...defaultProps} />)

    // Then
    expect(tree).toMatchSnapshot()
  })

  it('should render the list', () => {
    // Given
    const props = {
      ...defaultProps,
      historyList: [{ id: 1 }, { id: 2 }],
    }
    // When
    const tree = shallow(<HistoryList {...props} />)

    // Then
    expect(tree).toMatchSnapshot()
  })

  it('should call removeFromHistory on item icon click', () => {
    // Given
    const props = {
      ...defaultProps,
      historyList: [{ id: 1, name: 'yolo' }],
    }
    const tree = shallow(<HistoryList {...props} />)
    const item = tree.find(HistoryListItem).first()

    // When
    item.simulate('click')

    // Then
    expect(props.removeFromHistory).toHaveBeenCalled()
  })
})
