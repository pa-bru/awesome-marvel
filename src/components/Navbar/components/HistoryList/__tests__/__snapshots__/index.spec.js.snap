// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryList should render a specific message when the list is empty 1`] = `
<WithStyles(List)
  component="div"
  disablePadding={true}
  subheader={
    <WithStyles(ListSubheader)
      component="div"
    >
      Your History
    </WithStyles(ListSubheader)>
  }
>
  <WithStyles(ListItem)
    dense={true}
  >
    <WithStyles(ListItemText)
      primary="No History, start navigating to see your history"
    />
  </WithStyles(ListItem)>
</WithStyles(List)>
`;

exports[`HistoryList should render the error view 1`] = `
<ErrorView>
  Bad request
</ErrorView>
`;

exports[`HistoryList should render the list 1`] = `
<WithStyles(List)
  component="div"
  disablePadding={true}
  subheader={
    <WithStyles(ListSubheader)
      component="div"
    >
      Your History
    </WithStyles(ListSubheader)>
  }
>
  <React.Fragment>
    <withRouter(HistoryListItem)
      id={1}
      key="0"
      onClick={[Function]}
    />
    <withRouter(HistoryListItem)
      id={2}
      key="1"
      onClick={[Function]}
    />
    <WithStyles(Button)
      color="default"
      onClick={[Function]}
      variant="contained"
    >
      Clear History
    </WithStyles(Button)>
  </React.Fragment>
</WithStyles(List)>
`;

exports[`HistoryList should render the loading view 1`] = `<LoadingView />`;
